File: src/main/java/uk/co/reecedunn/intellij/plugin/core/lexer/CombinedLexer.java
/*
 * Copyright (C) 2016 Reece H. Dunn
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package uk.co.reecedunn.intellij.plugin.core.lexer;

import com.intellij.lexer.Lexer;
import com.intellij.lexer.LexerBase;
import com.intellij.psi.tree.IElementType;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.HashMap;
import java.util.Map;

public class CombinedLexer extends LexerBase {
    class State {
        final Lexer lexer;
        final int state;
        final int parentState;
        final int childState;
        final IElementType transition;

        State(Lexer lexer, int state, int parentState, int childState, IElementType transition) {
            this.lexer = lexer;
            this.state = state;
            this.parentState = parentState;
            this.childState = childState;
            this.transition = transition;
        }
    }

    private final Lexer mLanguage;
    private final Map<Integer, State> mStates = new HashMap<>();
    private final Map<IElementType, State> mTransitions = new HashMap<>();
    private int mStateMask;

    private Lexer mActiveLexer;
    private int mState;

    public CombinedLexer(Lexer language) {
        mLanguage = language;
        mStateMask = 0;
    }

    public void addState(Lexer lexer, int stateId, int parentStateId, IElementType transition) {
        addState(lexer, stateId, parentStateId, 0, transition);
    }

    public void addState(Lexer lexer, int stateId, int parentStateId, int childStateId, IElementType transition) {
        State state = new State(lexer, stateId, parentStateId, childStateId, transition);
        mStates.put(stateId, state);
        mTransitions.put(transition, state);
        mStateMask |= stateId;
    }

    // region Lexer

    @Override
    public final void start(@NotNull CharSequence buffer, int startOffset, int endOffset, int initialState) {
        mState = initialState & mStateMask;
        State state = mStates.getOrDefault(mState, null);
        if (state != null) {
            mActiveLexer = state.lexer;
            mLanguage.start(buffer, startOffset, endOffset, state.parentState);
            mActiveLexer.start(mLanguage.getBufferSequence(), mLanguage.getTokenStart(), mLanguage.getTokenEnd(), initialState & ~mStateMask);
        } else {
            mLanguage.start(buffer, startOffset, endOffset, initialState & ~mStateMask);
            state = mTransitions.getOrDefault(mLanguage.getTokenType(), null);
            if (state != null) {
                mActiveLexer = state.lexer;
                mActiveLexer.start(mLanguage.getBufferSequence(), mLanguage.getTokenStart(), mLanguage.getTokenEnd(), state.childState);
                mState = state.state;
            } else {
                mActiveLexer = mLanguage;
            }
        }
    }

    @Override
    public void advance() {
        if (mActiveLexer != mLanguage) {
            mActiveLexer.advance();
            if (mActiveLexer.getTokenType() == null) {
                mLanguage.advance();
                mState = 0;
                mActiveLexer = mLanguage;
            }
        } else {
            mLanguage.advance();
            State state = mTransitions.getOrDefault(mLanguage.getTokenType(), null);
            if (state != null) {
                mActiveLexer = state.lexer;
                mActiveLexer.start(mLanguage.getBufferSequence(), mLanguage.getTokenStart(), mLanguage.getTokenEnd(), state.childState);
                mState = state.state;
            }
        }
    }

    @Override
    public int getState() {
        return mActiveLexer.getState() | mState;
    }

    @Nullable
    @Override
    public IElementType getTokenType() {
        return mActiveLexer.getTokenType();
    }

    @Override
    public int getTokenStart() {
        return mActiveLexer.getTokenStart();
    }

    @Override
    public int getTokenEnd() {
        return mActiveLexer.getTokenEnd();
    }

    @NotNull
    @Override
    public CharSequence getBufferSequence() {
        return mLanguage.getBufferSequence();
    }

    @Override
    public int getBufferEnd() {
        return mLanguage.getBufferEnd();
    }

    // endregion
}

File: src/main/java/uk/co/reecedunn/intellij/plugin/core/lexer/CombinedLexer.kt
/*
 * Copyright (C) 2016-2018 Reece H. Dunn
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package uk.co.reecedunn.intellij.plugin.core.lexer

import com.intellij.lexer.Lexer
import com.intellij.lexer.LexerBase
import com.intellij.psi.tree.IElementType
import java.util.*

class CombinedLexer(private val mLanguage: Lexer) : LexerBase() {
    private val mStates = HashMap<Int, State>()
    private val mTransitions = HashMap<IElementType, State>()
    private var mStateMask: Int = 0

    private var mActiveLexer: Lexer? = null
    private var mState: Int = 0

    internal inner class State(val lexer: Lexer,
                               val state: Int,
                               val parentState: Int,
                               val childState: Int)

    fun addState(lexer: Lexer, stateId: Int, parentStateId: Int, transition: IElementType) {
        addState(lexer, stateId, parentStateId, 0, transition)
    }

    fun addState(lexer: Lexer, stateId: Int, parentStateId: Int, childStateId: Int, transition: IElementType) {
        val state = State(lexer, stateId, parentStateId, childStateId)
        mStates[stateId] = state
        mTransitions[transition] = state
        mStateMask = mStateMask or stateId
    }

    // region Lexer

    override fun start(buffer: CharSequence, startOffset: Int, endOffset: Int, initialState: Int) {
        mState = initialState and mStateMask
        var state: State? = mStates[mState]
        if (state != null) {
            mActiveLexer = state.lexer
            mLanguage.start(buffer, startOffset, endOffset, state.parentState)
            mActiveLexer!!.start(mLanguage.bufferSequence, mLanguage.tokenStart, mLanguage.tokenEnd, initialState and mStateMask.inv())
        } else {
            mLanguage.start(buffer, startOffset, endOffset, initialState and mStateMask.inv())
            state = mTransitions[mLanguage.tokenType]
            if (state != null) {
                mActiveLexer = state.lexer
                mActiveLexer!!.start(mLanguage.bufferSequence, mLanguage.tokenStart, mLanguage.tokenEnd, state.childState)
                mState = state.state
            } else {
                mActiveLexer = mLanguage
            }
        }
    }

    override fun advance() {
        if (mActiveLexer !== mLanguage) {
            mActiveLexer!!.advance()
            if (mActiveLexer!!.tokenType == null) {
                mLanguage.advance()
                val state = mTransitions[mLanguage.tokenType]
                if (state != null) {
                    mActiveLexer = state.lexer
                    mActiveLexer!!.start(mLanguage.bufferSequence, mLanguage.tokenStart, mLanguage.tokenEnd, state.childState)
                    mState = state.state
                } else {
                    mActiveLexer = mLanguage
                    mState = 0
                }
            }
        } else {
            mLanguage.advance()
            val state = mTransitions[mLanguage.tokenType]
            if (state != null) {
                mActiveLexer = state.lexer
                mActiveLexer!!.start(mLanguage.bufferSequence, mLanguage.tokenStart, mLanguage.tokenEnd, state.childState)
                mState = state.state
            }
        }
    }

    override fun getState(): Int = mActiveLexer!!.state or mState

    override fun getTokenType(): IElementType? = mActiveLexer!!.tokenType

    override fun getTokenStart(): Int = mActiveLexer!!.tokenStart

    override fun getTokenEnd(): Int = mActiveLexer!!.tokenEnd

    override fun getBufferSequence(): CharSequence = mLanguage.bufferSequence

    override fun getBufferEnd(): Int = mLanguage.bufferEnd

    // endregion
}
